File: backend/internal/compliance/compliance.go
package compliance

import (
	"time"
)

// Jurisdiction represents a legal territory with specific rules.
type Jurisdiction string

const (
	JurisdictionEU      Jurisdiction = "EU"
	JurisdictionGermany Jurisdiction = "DE"
	JurisdictionNL      Jurisdiction = "NL"
	// Add others as needed, e.g., JurisdictionUK
)

// ComplianceStatus indicates the result of an assessment.
type ComplianceStatus string

const (
	StatusMandatory ComplianceStatus = "MANDATORY"
	StatusVoluntary ComplianceStatus = "VOLUNTARY"
	StatusExempt    ComplianceStatus = "EXEMPT"
)

// ComplianceRequest captures the inputs required for a regulatory check.
type ComplianceRequest struct {
	Jurisdiction      Jurisdiction `json:"jurisdiction"`
	TotalITLoadKW     float64      `json:"totalItLoadKw"`     // Rated IT power
	PlanDate          time.Time    `json:"planDate"`          // When is the DC commissioning?
	HeatRecoveryReady bool         `json:"heatRecoveryReady"` // Is it already technically ready?

	// Feasibility Context (Optional - populated if assessment ran)
	DistanceToNetworkKm *float64 `json:"distanceToNetworkKm"`
	BestPaybackYears    *float64 `json:"bestPaybackYears"`
	HasHeatDemand       bool     `json:"hasHeatDemand"`
}

// RegulatoryReference identifies a specific provision in legislation.
type RegulatoryReference struct {
	Law     string `json:"law"`     // e.g. "Energieeffizienzgesetz (EnEfG)"
	Section string `json:"section"` // e.g. "§11(2)"
	Summary string `json:"summary"` // Short human-readable summary of the provision
	URL     string `json:"url"`     // Link to official source (EUR-Lex, BGBL, etc.)
}

// ComplianceResult is the output of the engine.
type ComplianceResult struct {
	Status             ComplianceStatus     `json:"status"`
	ApplicableLaw      string               `json:"applicableLaw"`
	ComplianceDeadline *time.Time           `json:"complianceDeadline,omitempty"`
	Reasoning          []string             `json:"reasoning"`
	RemediationSteps   []string             `json:"remediationSteps"`
	Citations          []RegulatoryReference `json:"citations"`
	Disclaimer         string               `json:"disclaimer"`
}

// RuleEngine defines the interface for jurisdiction-specific logic.
type RuleEngine interface {
	Check(req ComplianceRequest) ComplianceResult
}

// Engine is the entry point for compliance checks.
type Engine struct {
	rules map[Jurisdiction]RuleEngine
}

func NewEngine() *Engine {
	return &Engine{
		rules: map[Jurisdiction]RuleEngine{
			JurisdictionEU:      &EURules{},
			JurisdictionGermany: &GermanRules{},
			JurisdictionNL:      &DutchRules{},
		},
	}
}

// Evaluate determines the compliance status for the given request.
// It defaults to EU rules if the jurisdiction is not explicitly supported.
func (e *Engine) Evaluate(req ComplianceRequest) ComplianceResult {
	rule, ok := e.rules[req.Jurisdiction]
	if !ok {
		// Fallback or default to EU if in Europe?
		// For now, if unknown, we use EU as a baseline or return generic.
		// Let's use EU as baseline for "Other".
		rule = &EURules{}
	}
	return rule.Check(req)
}

File: frontend-next/src/lib/compliance.ts

export type ComplianceStatus = "MANDATORY" | "VOLUNTARY" | "EXEMPT";

export type Jurisdiction = "EU" | "DE" | "NL";

export interface ComplianceRequest {
    jurisdiction: Jurisdiction;
    totalItLoadKw: number;
    planDate: string; // ISO date string
    heatRecoveryReady: boolean;
}

export interface RegulatoryReference {
    law: string;
    section: string;
    summary: string;
    url: string;
}

export interface ComplianceResult {
    status: ComplianceStatus;
    applicableLaw: string;
    complianceDeadline?: string;
    reasoning: string[];
    remediationSteps: string[];
    citations: RegulatoryReference[];
    disclaimer: string;
}

export async function checkCompliance(req: ComplianceRequest): Promise<ComplianceResult> {
    const res = await fetch("/api/compliance/check", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(req),
    });

    if (!res.ok) {
        throw new Error("Failed to check compliance");
    }

    return res.json();
}

import { predictionService } from "./backend-client";
import { DataCenter } from "./schema";

export async function saveDataCenter(req: ComplianceRequest, name: string): Promise<{ dataCenter: DataCenter }> {
    const payload = {
        name: name,
        location: {
            latitude: 52.3676, // Default to Amsterdam for MVP if no geo input
            longitude: 4.9041
        },
        totalItLoadKw: req.totalItLoadKw,
        heatRecoveryEnabled: req.heatRecoveryReady,
        operatingHoursYear: 8760,
        dcType: "hyperscale", // Default
    };

    return predictionService.createDataCenter(payload);
}
